Better and more consistent error messages in torch.linalg

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62734

Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

Differential Revision: [D31823230](https://our.internmc.facebook.com/intern/diff/D31823230)

lezcano committed:
Better and more consistent error messages in torch.linalg

Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3078,8 +3070,8 @@ std::tuple<Tensor&, Tensor&, Tensor&> svd_out(const Tensor& self, bool some, boo
 
 std::tuple<Tensor, Tensor, Tensor> linalg_svd(const Tensor& self, bool full_matrices) {
-  TORCH_CHECK(self.dim() >= 2,
-              "svd input should have at least 2 dimensions, but has ", self.dim(), " dimensions instead");
+    TORCH_CHECK(self.dim() >= 2,
+                "svd input should have at least 2 dimensions, but has ", self.dim(), " dimensions instead");
 
     bool some = !full_matrices;
     Tensor U, S, V;
```
Indentation should be 2 spaces here.

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3081,13 +3073,12 @@ std::tuple<Tensor, Tensor, Tensor> linalg_svd(const Tensor& self, bool full_matr
-
-    Tensor Vh = V.conj().transpose(-2, -1);
-    return std::make_tuple(U, S, Vh);
+  bool some = !full_matrices;
+  Tensor U, S, V;
+  std::tie(U, S, V) = at::_svd_helper(self, some, /*compute_uv=*/true);
 
+  Tensor Vh = V.conj().transpose(-2, -1);
+  return std::make_tuple(U, S, Vh);
```
No idea why i tchose to go with this terrible diff, but this indents this paragraph to 2 spaces rather than 4, to be consistent with the rest of the file.

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -121,15 +122,15 @@ Tensor& linalg_det_out(const Tensor& self, Tensor& out) {
 Tensor linalg_det(const Tensor& self) {
-  squareCheckInputs(self);
+  squareCheckInputs(self, "linalg_det");
   TORCH_CHECK((at::isFloatingType(self.scalar_type()) || at::isComplexType(self.scalar_type())),
               "Expected a floating point or complex tensor as input");
```
I think these checks could be gathered into a single function again with a string argument for the function that is being checked.

Lezcano replied:
Done. Again, we should do a decent clean-up of linalg for 1.11, now that the functionality is sort of complete. I'm sure that while doing that we'll be able to make the code much easier to maintain and will really payoff in the future.

nikitaved commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -206,10 +206,11 @@ static inline void linearSolveCheckInputs(const Tensor& self, const Tensor& A, c
 // Validates input shapes for operations on batches of square matrices (inverse, cholesky, symeig, eig)
-static inline void squareCheckInputs(const Tensor& self) {
-  TORCH_CHECK(self.dim() >= 2, "Tensor of matrices must have at least 2 dimensions. ");
+static inline void squareCheckInputs(const Tensor& self, const char* const f_name) {
+  TORCH_CHECK(self.dim() >= 2, f_name, ": The input tensor must have at least 2 dimensions.");
```
nit: `f_name, "(): ..."`. Looks like this is a generic style convention for functions.

Lezcano replied:
Leaving as it is for now, as this is the format the other errors in linalg and advanced indexing follow.

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

nikitaved reviewed (approved):
LGTM! Thank you, Mario!

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

[ghstack-poisoned]

mruberry reviewed (approved):
I'm going to start the work of getting this stack landed, @Lezcano.

The clean-up looks good but would you update the names with their user-facing counterparts?

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31823230).

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

Differential Revision: [D31823230](https://our.internmc.facebook.com/intern/diff/D31823230)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31823230).

lezcano committed:
Update on "Better and more consistent error messages in torch.linalg"


Following https://github.com/pytorch/pytorch/pull/62715#discussion_r682610788
- squareCheckInputs takes a string with the name of the function
- We reuse more functions when checking the inputs

The state of the errors in torch.linalg is far from great though. We
leave a more comprehensive clean-up for the future.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

Differential Revision: [D31823230](https://our.internmc.facebook.com/intern/diff/D31823230)

[ghstack-poisoned]
